auth: start using SHA 512 hashing originated from musl with added yielding

andrzej-jackowski-scylladb opened the pull request:
This patch series contains the following changes:
 - Incorporation of `crypt_sha512.c` from musl to out codebase
 - Conversion of `crypt_sha512.c` to C++ and coroutinization
 - Coroutinization of `auth::passwords::check`
 - Enabling use of `__crypt_sha512` orignated from `crypt_sha512.c` for
   computing SHA 512 passwords of length <=255
 - Addition of yielding in the aforementioned hashing implementation.

The alien thread was a solution for reactor stalls caused by indivisible
password‑hashing tasks (https://github.com/scylladb/scylladb/issues/24524). 
However, because there is only one alien thread, overall hashing throughput was reduced
(see, e.g., https://github.com/scylladb/scylla-enterprise/issues/5711). To address this,
the alien‑thread solution is reverted, and a hashing implementation
with yielding is introduced in this patch series.

Before this patch series, ScyllaDB used SHA-512 hashing provided
by the `crypt_r` function, which in our case meant using the implementation
from the `libxcrypt` library. Adding yielding to this `libxcrypt`
implementation is problematic, both due to licensing (LGPL) and because the
implementation is split into many functions across multiple files. In
contrast, the SHA-512 implementation from `musl libc` has a more
permissive license and is concise, which makes it easier to incorporate
into the ScyllaDB codebase.

The performance of this solution was compared with the previous
implementation that used one alien thread and the implementation
after the alien thread was reverted. The results (median) of
`perf-cql-raw` with `--connection-per-request 1 --smp 10` parameters
are as follows:
 - Alien thread: 41.5 new connections/s per shard
 - Reverted alien thread: 244.1 new connections/s per shard
 - This commit (yielding in hashing): 198.4 new connections/s per shard

The roughly 20% performance deterioration compared to
the old implementation without the alien thread comes from the fact
that the new hashing algorithm implemented in `utils/crypt_sha512.cc`
performs an expensive self-verification and stack cleanup.

On the other hand, with smp=10 the current implementation achieves
roughly 5x higher throughput than the alien thread. In addition,
due to yielding added in this commit, the algorithm is expected
to provide similar protection from stalls as the alien thread did.
In a test that in parallel started a cassandra-stress workload and
created thousands of new connections using python-driver, the values of
`scylla_reactor_stalls_count` metric were as follows:
 - Alien thread: 109 stalls/shard total
 - Reverted alien thread: 13186 stalls/shard total
 - This commit (yielding in hashing): 149 stalls/shard total

Similarly, the `scylla_scheduler_time_spent_on_task_quota_violations_ms`
values were:
 - Alien thread: 1087 ms/shard total
 - Reverted alien thread: 72839 ms/shard total
 - This commit (yielding in hashing): 1623 ms/shard total

To summarize, yielding during hashing computations achieves similar
throughput to the old solution without the alien thread but also
prevents stalls similarly to the alien thread.

Fixes: scylladb/scylladb#26859
Refs: scylladb/scylla-enterprise#5711

No automatic backport. After this PR is completed, the alien thread should be rather reverted from older branches (2025.2-2025.4 because on 2025.1 it's already removed). Backporting of the other commits needs further discussion.

DarioMirovic commented on auth/passwords.hh:
```diff
@@ -80,6 +81,13 @@ sstring generate_salt(RandomNumberEngine& g, scheme scheme) {
 sstring hash_with_salt(const sstring& pass, const sstring& salt);
 
+///
+/// Newer version of `hash_with_salt` that returns a future.
```
"Newer" does not add anything of value. It states nothing about the function purpose, behavior, etc. Also, the other function is not deprecated here, and even then I'd rather put "deprecated" in the old function than "newer" in the new function. Consider using "Async" instead of "Newer".

DarioMirovic commented on test/boost/auth_passwords_test.cc:
```diff
@@ -97,6 +97,42 @@ SEASTAR_TEST_CASE(same_hashes_as_crypt_h) {
+    // The check below works because only the first 72 bytes are compared
+    BOOST_REQUIRE(h2_bcrypt == h3_bcrypt);
+    // The current implementation of bcrypt password hasing fails with passwords of length 512 and above
+    co_await try_and_expect_throw(too_long_password, salt_bcrypt);
```
You could use `BOOST_REQUIRE_THROW`:
```cpp
BOOST_REQUIRE_THROW(co_await auth::passwords::detail::hash_with_salt_async(too_long_password, salt_bcrypt);`
```

andrzej-jackowski-scylladb commented on auth/passwords.cc:
```diff
@@ -21,27 +23,46 @@ static thread_local crypt_data tlcrypt = {};
 
+seastar::future<sstring> hash_with_salt_async(const sstring& pass, const sstring& salt) {
+    const char * res;
+    // Only SHA-512 hashes for passphrases shorter than 256 bytes can be computed using
```
I agree, but I wanted to prevent a situation that a customer that already uses a password longer than 256 bytes cannot login to the system after an upgrade.

I reported https://github.com/scylladb/scylladb/issues/26842 to discuss what limit we should have, document it, and enforce it for new passwords in the future.

andrzej-jackowski-scylladb replied:
@mykaul I'm concerned we cannot really prevent the upgrade. The passwords kept persistently are hashed, and the hashes have the same length. The only moments when we know the password length are: 1) creating a new password (in CREATE/ALTER) and 2) authentication (on each new connection). At upgrade time we don't know the lengths of the passwords used in the system unless we start storing the password length somewhere (storing the exact length probably introduces a security vulnerability, but maybe we can just store a boolean indicating whether the password is over 256 bytes or not).

So for now I've kept the fallback to the old algorithm in case the password is too long. The fallback is needed anyway, because hashes created with algorithms other than SHA-512 can be injected (e.g., using `2a`=`bcrypt` in `CREATE ROLE ... WITH HASHED PASSWORD ...`). Example:
https://github.com/scylladb/scylladb/blob/d6c64097ad163905bd3b9894152ca91d1244c481/test/cqlpy/test_describe.py#L1664)

Similarly, `CREATE ROLE ... WITH HASHED PASSWORD ...` allows injecting SHA-512 hashes created from passwords longer than 256 bytes, even after such an upgrade.

Therefore, I think the first step should be documenting the limit and deprecating longer passwords.

andrzej-jackowski-scylladb commented on auth/passwords.cc:
```diff
@@ -21,24 +21,28 @@ static thread_local crypt_data tlcrypt = {};
-
-    if (e && (e[0] != '*')) {
-        return;
+    try {
```
Yes, `verify_scheme` starts to throw twice. On the other hand:
1) `verify_scheme` used only one, during the system start
2) Throwing in `verify_scheme` is extremely unlikely. It means our system library doesn't support SHA-512. Probably it can happen only if our build system is broken, someone compiles ScyllaDB on a very peculiar OS, or in some very distant future. In either case, throwing the exception twice has no impact.

nuivall commented on auth/passwords.cc:
```diff
@@ -21,27 +23,46 @@ static thread_local crypt_data tlcrypt = {};
+    try {
+        verify_hashing_output(e);
+    }
+    catch (std::system_error& ex) {
```
style: usually (always?) we put catch in the same line as }

nuivall commented on auth/passwords.cc:
```diff
@@ -49,7 +49,18 @@ sstring hash_with_salt(const sstring& pass, const sstring& salt) {
+    // crypt_r implementation from `<crypt.h>`, which can stall.
+    if (salt.starts_with(prefix_for_scheme(scheme::sha_512)) && pass.size() <= 255) {
+        char buf[128];
+        res = co_await __crypt_sha512(pass.c_str(), salt.c_str(), buf);
```
is it by any chance res pointing to data from buf?

nuivall reviewed (changes requested):
left some feedback, minor, except buf question

andrzej-jackowski-scylladb commented on auth/passwords.cc:
```diff
@@ -21,27 +23,46 @@ static thread_local crypt_data tlcrypt = {};
+    try {
+        verify_hashing_output(e);
+    }
+    catch (std::system_error& ex) {
```
fixed

andrzej-jackowski-scylladb commented:
> * [test_distributed_count_all](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21738/testReport/junit/distributed_queries_test/TestDistributedAggregations/release___dtest_with_tablets___test_distributed_count_all) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_distributed_count_all)

Looks like https://github.com/scylladb/scylladb/issues/16709

dani-tweig commented:
Opened a bug: https://scylladb.atlassian.net/browse/PM-65

andrzej-jackowski-scylladb commented on test/boost/auth_passwords_test.cc:
```diff
@@ -16,15 +16,21 @@
 
 #include "seastarx.hh"
 
+extern "C" {
```
I added it here to be consistent with the existing code in `auth/password.cc` and `auth/default_authorizer.cc` as both contain:
```
extern "C" {
#include <crypt.h>
#include <unistd.h>
}
```

In my environment both `crypt.h` and `unistd.h` have `__BEGIN_DECLS`, therefore I think the `extern "C"` is redundant. However, I think it's better to test the same code (including the redundant `extern "C"`) as the product uses.

andrzej-jackowski-scylladb commented on auth/passwords.cc:
```diff
@@ -49,7 +49,18 @@ sstring hash_with_salt(const sstring& pass, const sstring& salt) {
+    // Only SHA-512 hashes for passphrases shorter than 256 bytes can be computed using
+    // the __crypt_sha512 method. For other computations, we fall back to the
+    // crypt_r implementation from `<crypt.h>`, which can stall.
+    if (salt.starts_with(prefix_for_scheme(scheme::sha_512)) && pass.size() <= 255) {
```
The code was already improved in https://github.com/scylladb/scylladb/pull/21749, I don't think we should modify it further.

andrzej-jackowski-scylladb commented:
v5:
 - Added relevant information to `licenses/README.md`
 - Modified wording in `utils/crypt_sha512.cc` comments
 - Fixed `licenses/musl-icense.txt` name to `licenses/musl-license.txt`
 - Updated commit messages
 - Added `deprecated` to `hash_with_salt`
 - Added missing comment to `test/boost/auth_passwords_test.cc`

andrzej-jackowski-scylladb commented on test/boost/auth_passwords_test.cc:
```diff
@@ -97,6 +97,32 @@ SEASTAR_TEST_CASE(same_hashes_as_crypt_h) {
+    auto h2_bcrypt = co_await auth::passwords::detail::hash_with_salt_async(p2, salt_bcrypt);
+    auto h3_bcrypt = co_await auth::passwords::detail::hash_with_salt_async(p3, salt_bcrypt);
+    BOOST_REQUIRE(h1_bcrypt != h2_bcrypt);
+    // The check below works because only the first 72 bytes are compared
```
Added comment to `test/boost/auth_passwords_test.cc` to address the first paragraph.

andrzej-jackowski-scylladb commented on auth/passwords.hh:
```diff
@@ -75,11 +76,19 @@ sstring generate_salt(RandomNumberEngine& g, scheme scheme) {
 
 ///
 /// Hash a password combined with an implementation-specific salt string.
+/// Deprecated in favor of `hash_with_salt_async`.
```
I added `deprecated` to `hash_with_salt`, as well as explanation to e9ce482ea0ab44bc25faa1bd69cd9467549dee10:

```
The old method, `hash_with_salt`, is marked as deprecated because
new code should use the new `hash_with_salt_async` function.
We are not removing `hash_with_salt` now to reduce the regression risk
of changing the hashing implementation—at least the methods that change
persistent hashes (CREATE, ALTER) will continue to use the old hashing
method. However, in the future, `hash_with_salt` should be entirely
removed.
```

andrzej-jackowski-scylladb commented on auth/allow_all_authenticator.cc:
```diff
@@ -22,7 +21,6 @@ static const class_registrator<
         ::service::raft_group0_client&,
-        ::service::migration_manager&,
-        utils::alien_worker&> registration("org.apache.cassandra.auth.AllowAllAuthenticator");
+        ::service::migration_manager&> registration("org.apache.cassandra.auth.AllowAllAuthenticator");
```
I agree, added the following paragraph to 78a8f5d0475067a91959b3437821363b237f3187:
```
The alien thread was a solution for reactor stalls caused by indivisible
password‑hashing tasks (https://github.com/scylladb/scylladb/issues/24524). However, because
there is only one alien thread, overall hashing throughput was reduced
(see, e.g., https://github.com/scylladb/scylla-enterprise/issues/5711). To address this,
the alien‑thread solution is reverted, and a hashing implementation
with yielding will be introduced later in this patch series.
```

andrzej-jackowski-scylladb commented on utils/crypt_sha512.c:
```diff
@@ -0,0 +1,371 @@
+/*
+ * public domain sha512 crypt implementation
```
Thank you for spotting the missing `l` in `licenses/musl-icense.txt`, I fixed it.

Regarding the musl license I updated the patches to follow the Avi's guideline:
```
1. Copy the file as is and don't delete any licensing information (only add to it)
2. Record anything relevant (including that it was taking from musl which is under overall MIT license) in licenses/.
```

I modified the comment in `utils/crypt_sha512.cc` to not modify any existing licensing lines.
I also added all relevant information to `licenses/README.md`

andrzej-jackowski-scylladb commented on utils/crypt_sha512.cc:
```diff
@@ -1,20 +1,31 @@
+ * Copyright (C) 2025-present ScyllaDB
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ *
+ * Musl as a whole is published under the MIT license and this particular file
```
Replied in https://github.com/scylladb/scylladb/pull/26860#discussion_r2582259476

andrzej-jackowski-scylladb commented:
> There are conflicts that needs to be resolved before running the CI again.

Yes, I waited for https://github.com/scylladb/scylladb/pull/27434 to be merged, because I knew it will generate more conflicts. Will update this PR today.

nyh commented on auth/passwords.cc:
```diff
@@ -61,8 +81,9 @@ no_supported_schemes::no_supported_schemes()
 
-bool check(const sstring& pass, const sstring& salted_hash) {
-    return detail::hash_with_salt(pass, salted_hash) == salted_hash;
+seastar::future<bool> check(const sstring& pass, const sstring& salted_hash) {
```
Again, nitpick: In Scylla we usually don't write seastar::future, we just write "future".

nyh commented on auth/passwords.hh:
```diff
@@ -75,10 +76,18 @@ sstring generate_salt(RandomNumberEngine& g, scheme scheme) {
 /// \throws \ref std::system_error when an unexpected implementation-specific error occurs.
 ///
-sstring hash_with_salt(const sstring& pass, const sstring& salt);
+[[deprecated("Use hash_with_salt_async instead")]] sstring hash_with_salt(const sstring& pass, const sstring& salt);
```
I'm not convinced the argument to deprecate this function instead of removing it is valid: I don't understand how it helps "reduce regression risk" if we keep this deprecated function to be used in half of the cases - but still not all of them.

I'm worried that it's even the opposite: Because in some of the cases (?) we still use the old function, it's possible that the new function has a bug but a test we have that might have caught this bug happens to be one that still uses the old function.

nyh commented on auth/passwords.hh:
```diff
@@ -84,6 +84,7 @@ sstring generate_salt(RandomNumberEngine& g, scheme scheme) {
 
 ///
 /// Async version of `hash_with_salt` that returns a future.
+/// If possible, hashing uses `coroutine::maybe_yield` to prevent reactor stalls.
```
Why "if possible"? You do use it. I would say something like:
This function avoids stalls by yielding during long computations.
(or use whatever wording you prefer).

nyh commented on test/boost/auth_passwords_test.cc:
```diff
@@ -58,6 +58,45 @@ SEASTAR_TEST_CASE(correct_passwords_authenticate) {
+
+    for (const char* p : passwords) {
+        auto res = co_await auth::passwords::detail::hash_with_salt_async(p, salt);
+        BOOST_REQUIRE(res == auth::passwords::detail::hash_with_salt(p, salt));
```
It's good to have this test today, but if I understand correctly, the intention is to delete the old hash_with_salt(), so this test will have to be removed.
Perhaps it will be better (now, or later) to open-code hash_with_salt() here to call the original library function and compares the new one to it, so that even after we remove the old hash_with_salt() this test will be able to still run.

nyh reviewed (approved):
I did have some minor comments, but in general I think this version is good enough to be merged and if needed, improved later by further patches.
